### PR TITLE
[PS-248] Feature/add connections enabled endpoint

### PR DESCRIPTION
**Pull request by MGibson1**
## Type of change
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective
Moves Organization models into subfolders and sub-namespaces.

API endpoints for Organization Connection CRUD operations.

## Code changes
Most files are changed as a result of the model move.

* **OrganizationConnectionController**: CRUD operations for oragnization connections
  * create, update take a general version of the organization connection request and convert to a typed version depending on the type enum in the request.
  * `getConnectionEnabled` is used to determine whether billing sync settings are displayed in client.
* **OrganizatonConnectionRequestModel**: Has two types, one normal and one generic. The normal one is bound to the endpoint and the generic is used for automatic type checking once we know the organization connection type. This keeps us from having to deal with config strings beyond the controller.
* **OrganizationConnectionResponse**: response when getting and organization connection
  * null Id is used to denote that no connection was found to the clients. This keeps us from having to handle null responses or matching a `0` guid
  * `Config` is of type `object` so we can use our typical response handling logic to parse it into a first-class object rather than a serialized string client-side.
* **OrganizationConnection**: Type parameter overload since you can't pass type objects as generic parameters
* **OrgnaizationConnectionData**: Intermediate type between response/request and etity.
* **{create,update,delete}OragnizationConnectionCommand**: simple crud commands for organization connections
* **OrgnaizationConnectionRespository**: Gets enabled by c# filtering rather than on Db. We can only ever have one per org right now anyway...
* **DB changes**: fixes some broken sprocs for createing organization connection and changes the get by organization id sproc to get all and filter them in c#

### Testing changes
* **BitAutoDataAttribute**: Added data detector that skips pre-loading of data. This is recommended for randomly generated theory paramters, like through autofixture.
*  Good coverage of commands and controller.

## Testing requirements
None. Integration testing will be done with web client, but functionality is unit tested.

## Before you submit
- [x] I have checked for formatting errors (`dotnet tool run dotnet-format --check`) (required)
- [x] If making database changes - I have also updated Entity Framework queries and/or migrations
- [x] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

File: test/Api.Test/AutoFixture/Attributes/ControllerCustomizeAttribute.cs [hunk 1]
```diff
@@ -4,6 +4,11 @@
 
 namespace Bit.Api.Test.AutoFixture.Attributes
 {
+    /// <summary>
+    /// Disables setting of Auto Properties on the Controller to avoid ASP.net initialization errors from a mock environment. Still sets constructor dependencies.
+    /// </summary>
+    /// <param name="controllerType">The Type of the controller to allow autofixture to create</param>
```
Comment (justindbaur) [inline test/Api.Test/AutoFixture/Attributes/ControllerCustomizeAttribute.cs:10]: nit: A `param` comment shouldn't go on the class, this should be on the constructor.

File: test/Common/Helpers/AssertHelper.cs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
 using System.Linq;
+using System.Linq.Expressions;
```
Comment (justindbaur) [inline test/Common/Helpers/AssertHelper.cs:3]: My favorite namespace 😄

File: src/Core/Enums/OrganizationConnectionType.cs [hunk 3]
```diff
@@ -2,6 +2,6 @@
 {
     public enum OrganizationConnectionType : byte
     {
-        CloudBillingSync,
+        CloudBillingSync = 1,
```
[the remark this answers is not in the document]
Comment (MGibson1) [reply src/Core/Enums/OrganizationConnectionType.cs:5]: It's a bad idea to use the default value to encode something with meaning. think of it like this: if we had a default, that'd be zero, but we don't and there's no reason to think that billing sync will be default in the long run

File: src/Core/Models/Data/Organizations/OrganizationConnections/OrganizationConnectionData.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+using System;
+using Bit.Core.Entities;
+using Bit.Core.Enums;
+
+namespace Bit.Core.Models.Data.Organizations.OrganizationConnections
+{
+    public class OrganizationConnectionData<T> where T : new()
+    {
+        public Guid? Id { get; set; }
```
[the remark this answers is not in the document]
Comment (MGibson1) [reply src/Core/Models/Data/Organizations/OrganizationConnections/OrganizationConnectionData.cs:11]: When a connection is requested, but doesn't exist. Our other option is to return a null, but then we need to handle it as an error, which didn't feel right.

File: test/Common/Helpers/AssertHelper.cs (hunk 2 above)
Comment (MGibson1) [reply test/Common/Helpers/AssertHelper.cs:3]: I thought you'd like that. I'm actually pretty happy with the expression, too. I'm using it to make a matcher for NSubstitute that turned out to be pretty useful for this set of tests

File: src/Api/Models/Request/Organizations/OrganizationConnectionRequestModel.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+using Bit.Core.Utilities;
+
+namespace Bit.Api.Models.Request.Organizations
+{
+    public class OrganizationConnectionRequestModel
+    {
+        public OrganizationConnectionType Type { get; set; }
+        public Guid OrganizationId { get; set; }
+        public bool Enabled { get; set; }
+        public string Config { get; set; }
```
[the remark this answers is not in the document]
Comment (MGibson1) [reply src/Api/Models/Request/Organizations/OrganizationConnectionRequestModel.cs:15]: nice idea!

File: test/Api.Test/AutoFixture/Attributes/ControllerCustomizeAttribute.cs (hunk 1 above)
Comment (MGibson1) [reply test/Api.Test/AutoFixture/Attributes/ControllerCustomizeAttribute.cs:10]: Fair enough, that's what I get for blindly copying documentation over from another method

File: test/Common/Helpers/AssertHelper.cs (hunk 2 above)
Comment (justindbaur) [reply test/Common/Helpers/AssertHelper.cs:3]: Yeah they look really good.

**Review by justindbaur — APPROVED** [review]
Looks great!

**Pull request merged**